File: Ahorn/entities/vvvvvvGravityController.jl
# Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
# See the LICENCE file in the repository root for full licence text.

module GravityHelperVvvvvvGravityController

using ..Ahorn, Maple

const PLUGIN_VERSION = "1"

const default_vvvvvv_sound = "event:/gravityhelper/toggle"

@mapdef Entity "GravityHelper/VvvvvvGravityController" VvvvvvGravityController(
    x::Integer, y::Integer,
    pluginVersion::String=PLUGIN_VERSION,
    persistent::Bool=true,
    mode::Integer=2,
    disableGrab::Bool=true,
    disableDash::Bool=true,
    flipSound::String=default_vvvvvv_sound
)

const placements = Ahorn.PlacementDict(
    "VVVVVV Gravity Controller (GravityHelper)" => Ahorn.EntityPlacement(
        VvvvvvGravityController,
    ),
)

const vvvvvvModes = Dict{String, Integer}(
    "Trigger-based" => 0,
    "Off" => 1,
    "On" => 2,
)

Ahorn.editingIgnored(entity::VvvvvvGravityController, multiple::Bool=false) = String["modVersion", "pluginVersion"]

Ahorn.editingOptions(entity::VvvvvvGravityController) = Dict{String, Any}(
    "mode" => vvvvvvModes
)

const sprite = "objects/GravityHelper/gravityController/circle"
const sprite_dot = "objects/GravityHelper/gravityController/circle_dot"
const sprite_spikes = "objects/GravityHelper/gravityController/spikes"

function Ahorn.selection(entity::VvvvvvGravityController)
    x, y = Ahorn.position(entity)
    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::VvvvvvGravityController)
    icon = get(entity.data, "persistent", false) ? sprite_dot : sprite
    Ahorn.drawSprite(ctx, icon, 0, 0)
    Ahorn.drawSprite(ctx, sprite_spikes, 0, 0)
end

end

File: Loenn/entities/vvvvvvGravityController.lua
-- Copyright (c) Shane Woolcock. Licensed under the MIT Licence.
-- See the LICENCE file in the repository root for full licence text.

local consts = require("mods").requireFromPlugin("consts")
local helpers = require("mods").requireFromPlugin("helpers")
local drawableSprite = require("structs.drawable_sprite")

local placementData = helpers.createPlacementData('1', {
    persistent = true,
    mode = 2, -- On
    flipSound = "event:/gravityhelper/toggle",
    disableGrab = true,
    disableDash = true,
    disableWallJump = true,
    solidTilesBehavior = "Flip",
    otherPlatformBehavior = "Flip",
    extraJumpsBehavior = "Flip",
})

local vvvvvvGravityController = {
    name = "GravityHelper/VvvvvvGravityController",
    depth = -8500,
    ignoredFields = consts.ignoredFields,
    fieldOrder = {
        "x", "y",
        "flipSound", "mode",
        "solidTilesBehavior", "otherPlatformBehavior",
        "extraJumpsBehavior", "disableDash", "disableGrab",
        "disableWallJump", "persistent"
    },
    fieldInformation = {
        mode = consts.fieldInformation.vvvvvvMode,
        solidTilesBehavior = {
            editable = false,
            options = {
                "Flip",
                "Jump",
                "None",
            }
        },
        otherPlatformBehavior = {
            editable = false,
            options = {
                "Flip",
                "Jump",
                "None",
            }
        },
        extraJumpsBehavior = {
            editable = false,
            options = {
                "Flip",
                "Jump",
                "None",
            }
        },
    },
    placements = {
        {
            name = "single_room",
            data = helpers.union(placementData, {
                persistent = false,
            }),
        },
        {
            name = "persistent",
            data = helpers.union(placementData),
        },
    },
}

function vvvvvvGravityController.sprite(room, entity)
    local spriteName = entity.persistent and "objects/GravityHelper/gravityController/circle_dot" or "objects/GravityHelper/gravityController/circle"
    local iconSprite = drawableSprite.fromTexture(spriteName, entity)
    local typeSprite = drawableSprite.fromTexture("objects/GravityHelper/gravityController/spikes", entity)
    return {iconSprite, typeSprite}
end

return vvvvvvGravityController
